### ISSUE TITLE
You can use the following code to make the Diff transformer not OOM.

### ISSUE THREAD

**Issue by xumingyu2021** [erogol/BlaGPT#1]
from flash_attn import flash_attn_func

        bsz, tgt_len, embed_dim = x.size()
        src_len = tgt_len

        q = self.q_proj(x)
        k = self.k_proj(x)
        v = self.v_proj(x)

        q = q.view(bsz, tgt_len, 2 * self.num_heads, self.head_dim)
        k = k.view(bsz, src_len, 2 * self.num_kv_heads, self.head_dim)
        v = v.view(bsz, src_len, self.num_kv_heads, 2, self.head_dim)

        cos, sin = self.rotary(q)
        
        q = apply_rotary_emb(q, cos, sin)
        k = apply_rotary_emb(k, cos, sin)

        q = q.reshape(bsz, tgt_len, self.num_heads, 2, self.head_dim)
        k = k.reshape(bsz, src_len, self.num_kv_heads, 2, self.head_dim)
        q1, q2 = q[:, :, :, 0], q[:, :, :, 1]
        k1, k2 = k[:, :, :, 0], k[:, :, :, 1]
        v1, v2 = v[:, :, :, 0], v[:, :, :, 1]

        attn11 = flash_attn_func(q1, k1, v1, causal=True)
        attn12 = flash_attn_func(q1, k1, v2, causal=True)
        attn1 = torch.cat([attn11, attn12], dim=-1)
        
        attn21 = flash_attn_func(q2, k2, v1, causal=True)
        attn22 = flash_attn_func(q2, k2, v2, causal=True)
        attn2 = torch.cat([attn21, attn22], dim=-1)
        
        lambda_1 = torch.exp(torch.sum(self.lambda_q1 * self.lambda_k1, dim=-1).float()).type_as(q)
        lambda_2 = torch.exp(torch.sum(self.lambda_q2 * self.lambda_k2, dim=-1).float()).type_as(q)
        lambda_full = lambda_1 - lambda_2 + self.lambda_init
        attn = attn1 - lambda_full * attn2

        attn = self.subln(attn)
        attn = attn * (1 - self.lambda_init)
        attn = attn.reshape(bsz, tgt_len, self.num_heads * 2 * self.head_dim)
        
        attn = self.out_proj(attn)
        return attn

**Comment by xumingyu2021** [comment]
**And the following is the log of Diff transformer**
====================================================================================================
import datetime
import os
import sys

from coqpit import Coqpit
from optimizers import get_optimizer
from utils import get_model

with open(sys.argv[0]) as f:
    code = f.read()  # read the code of this file ASAP, for logging
import glob
import time
from dataclasses import dataclass, field

import numpy as np
import torch
import torch.distributed as dist
from torch.nn.parallel import DistributedDataParallel as DDP

os.environ["TORCH_NCCL_ASYNC_ERROR_HANDLING"] = "1"
os.environ["NCCL_TIMEOUT"] = "1800"

torch._dynamo.config.optimize_ddp = False
torch.set_float32_matmul_precision("high")

....

step:5080/5100 lr:1e-05 train_loss:3.3226 train_time:3018113ms step_avg:595.29ms
step:5081/5100 lr:9.5e-06 train_loss:3.6045 train_time:3018697ms step_avg:595.29ms
step:5082/5100 lr:9e-06 train_loss:3.4275 train_time:3019280ms step_avg:595.28ms
step:5083/5100 lr:8.500000000000002e-06 train_loss:3.5455 train_time:3019865ms step_avg:595.28ms
step:5084/5100 lr:8e-06 train_loss:3.3673 train_time:3020449ms step_avg:595.28ms
step:5085/5100 lr:7.5e-06 train_loss:3.4843 train_time:3021035ms step_avg:595.28ms
step:5086/5100 lr:7e-06 train_loss:3.5416 train_time:3021620ms step_avg:595.28ms
step:5087/5100 lr:6.5e-06 train_loss:3.4344 train_time:3022204ms step_avg:595.27ms
step:5088/5100 lr:6e-06 train_loss:3.3356 train_time:3022789ms step_avg:595.27ms
step:5089/5100 lr:5.5e-06 train_loss:3.2701 train_time:3023372ms step_avg:595.27ms
step:5090/5100 lr:5e-06 train_loss:3.4307 train_time:3023956ms step_avg:595.27ms
step:5091/5100 lr:4.5e-06 train_loss:3.6002 train_time:3024541ms step_avg:595.26ms
step:5092/5100 lr:4e-06 train_loss:3.5341 train_time:3025127ms step_avg:595.26ms
step:5093/5100 lr:3.5e-06 train_loss:3.3900 train_time:3025712ms step_avg:595.26ms
step:5094/5100 lr:3e-06 train_loss:3.4639 train_time:3026298ms step_avg:595.26ms
step:5095/5100 lr:2.5e-06 train_loss:3.2523 train_time:3026884ms step_avg:595.26ms
step:5096/5100 lr:2e-06 train_loss:3.3305 train_time:3027470ms step_avg:595.26ms
step:5097/5100 lr:1.5e-06 train_loss:3.2482 train_time:3028057ms step_avg:595.25ms
step:5098/5100 lr:1e-06 train_loss:3.5210 train_time:3028643ms step_avg:595.25ms
step:5099/5100 lr:5e-07 train_loss:3.4312 train_time:3029230ms step_avg:595.25ms
step:5100/5100 lr:0.0 train_loss:3.3198 train_time:3029817ms step_avg:595.25ms
step:5100/5100 val_loss:3.3601 train_time:3029880ms step_avg:595.26ms
step:5100/5100 val_loss:3.3601 train_time:3029880ms step_avg:595.26ms
peak memory consumption: 42072 MiB

**Comment by erogol** [comment]
cool thanks.
